pkg/controller/controller_utils.go

```diff
@@ -0,0 +1,281 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api/validation"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client/cache"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client/record"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util/wait"
+	"github.com/golang/glog"
+	"sync/atomic"
+	"time"
+)
+
+const CreatedByAnnotation = "kubernetes.io/created-by"
+
+// Expectations are a way for replication controllers to tell the rc manager what they expect. eg:
+//	RCExpectations: {
+//		rc1: expects  2 adds in 2 minutes
+//		rc2: expects  2 dels in 2 minutes
+//		rc3: expects -1 adds in 2 minutes => rc3's expectations have already been met
+//	}
+//
+// Implementation:
+//	PodExpectation = pair of atomic counters to track pod creation/deletion
+//	RCExpectationsStore = TTLStore + a PodExpectation per rc
+//
+// * Once set expectations can only be lowered
+// * An RC isn't synced till its expectations are either fulfilled, or expire
+// * Rcs that don't set expectations will get woken up for every matching pod
+
+// expKeyFunc to parse out the key from a PodExpectation
+var expKeyFunc = func(obj interface{}) (string, error) {
+	if e, ok := obj.(*PodExpectations); ok {
+		return e.key, nil
+	}
+	return "", fmt.Errorf("Could not find key for obj %#v", obj)
+}
+
+// RCExpectations is a ttl cache mapping rcs to what they expect to see before being woken up for a sync.
+type RCExpectations struct {
+	cache.Store
+}
+
+// GetExpectations returns the PodExpectations of the given rc.
+func (r *RCExpectations) GetExpectations(rc *api.ReplicationController) (*PodExpectations, bool, error) {
+	rcKey, err := rcKeyFunc(rc)
+	if err != nil {
+		return nil, false, err
+	}
+	if podExp, exists, err := r.GetByKey(rcKey); err == nil && exists {
+		return podExp.(*PodExpectations), true, nil
+	} else {
+		return nil, false, err
+	}
+}
+
+// SatisfiedExpectations returns true if the replication manager has observed the required adds/dels
+// for the given rc. Add/del counts are established by the rc at sync time, and updated as pods
+// are observed by the replication manager's podController.
+func (r *RCExpectations) SatisfiedExpectations(rc *api.ReplicationController) bool {
+	if podExp, exists, err := r.GetExpectations(rc); exists {
+		if podExp.Fulfilled() {
```

Comment:
Judging from your next commend you don't want me to do this right? Since we initialize it to the expected value before creating pods, and only decrement them from pods events or if a create/delete fails. Or did you have another race condition?